### PR TITLE
Fixed Issue #168: Prevent Users from Rejoining Events They Are Already Part Of

### PR DESCRIPTION
**Pull request by leagrieder**
Problem:
We encountered an issue where users could rejoin events they were already part of, leading to redundancy and potential inconsistencies in our event management logic. Specifically:

- Guests could join the same event multiple times.
- Staff members could re-register as guests for events where they were already listed as such.
- Event organizers could attempt to join events they were organizing

Solution:
To address this, I implemented a preliminary role check before the execution of the joinEvent function. The system now verifies the user's current role within the event context to determine if the operation should proceed. This check ensures:
- Guests, staff, and organizers are blocked from joining an event if they are already participating in their respective capacities.

I am still trying to understand why when we click on "join event" from the map and we are not a guest yet, once we successfully joined the event if I click again directly on the button it joins the event again... I need to go back to the Home Screen and go to the map again for the update to be active for some reason - UPDATE: FIXED

### CODE REVIEW

**Comment by arnaudlvq** [comment]
### Update of the Events List

The issue in the `FindEventMapScreen`, where users feel like they joined events twice, is due to the untouched local state in the `findEventViewModel`.

To resolve this:
1. When the `joinEvent` function is called, update the local list of events.
2. Ensure the `getRole` method returns the updated role of the user.

Great job on fixing the server-side issue.

File: app/src/androidTest/java/com/monkeyteam/chimpagne/FindEventScreenTest.kt [hunk 1]
```diff
@@ -91,7 +107,50 @@ class FindEventScreenTest {
...
+    composeTestRule.setContent {
+      val navController = rememberNavController()
+      val navActions = NavigationActions(navController)
+
+      FindEventMapScreen({}, findViewModel, accountViewModel, navActions)
+    }
+
+    accountManager.signInTo(anAccount)
+    val n = sampleEvent.getRole(anAccount.firebaseAuthUID)
+    if (n == ChimpagneRole.GUEST) {
```
Comment (Athanaze) [inline app/src/androidTest/java/com/monkeyteam/chimpagne/FindEventScreenTest.kt:135]: A when should be used here:
https://kotlinlang.org/docs/control-flow.html#when-expression

Comment (leagrieder) [reply app/src/androidTest/java/com/monkeyteam/chimpagne/FindEventScreenTest.kt]: Deleted the logs, they were not necessary, I had forgotten to remove them :)

File: app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt [hunk 2]
```diff
@@ -386,12 +400,46 @@ fun FindEventMapScreen(
...
+      // Check if the user is not logged in
+      !accountViewModel.isUserLoggedIn() -> {
+        // Redirect user to login screen
+        showPromptLogin = true
+      }
+
+      // The user has not yet joined the event
+      currentEvent?.getRole(accountUIState.currentUserAccount?.firebaseAuthUID ?: "") ==
+          ChimpagneRole.NOT_IN_EVENT -> {
+        Log.d("m", "BLOCK 1")
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt:413]: Unclear log message

File: app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt [hunk 3]
```diff
@@ -386,12 +400,46 @@ fun FindEventMapScreen(
...
+              { Toast.makeText(context, stringResFailiure, Toast.LENGTH_SHORT).show() })
+
+          navObject.clearAndNavigateTo(Route.VIEW_DETAIL_EVENT_SCREEN + "/${event.id})", false)
+        }
+      }
+
+      // The user has already joined the event, or is a staff for the event,or is the organizer
+      currentEvent?.getRole(accountUIState.currentUserAccount?.firebaseAuthUID ?: "") ==
+          ChimpagneRole.STAFF -> {
+        Log.d("m", "BLOCK 2")
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt:430]: Unclear log message

File: app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt [hunk 4]
```diff
@@ -386,12 +400,46 @@ fun FindEventMapScreen(
...
+        Toast.makeText(context, stringResStaff, Toast.LENGTH_SHORT).show()
+      }
+      currentEvent?.getRole(accountUIState.currentUserAccount?.firebaseAuthUID ?: "") ==
+          ChimpagneRole.OWNER -> {
+        Log.d("m", "BLOCK 2")
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt:435]: Unclear log message

File: app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt [hunk 5]
```diff
@@ -386,12 +400,46 @@ fun FindEventMapScreen(
...
+        Toast.makeText(context, stringResOwner, Toast.LENGTH_SHORT).show()
+      }
+      currentEvent?.getRole(accountUIState.currentUserAccount?.firebaseAuthUID ?: "") ==
+          ChimpagneRole.GUEST -> {
+        Log.d("m", "BLOCK 3")
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/FindEventScreen.kt:440]: Unclear log message

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -84,6 +85,11 @@
     <string name="links_to_social_media">Add your social media groups</string>
     <string name="my_events_screen_past_events_list_name">Past events</string>
     <string name="my_events_screen_empty_past_events_list">There is not past events yet</string>
+    <string name="join_event_failiure">Failed to Join the Event</string>
+    <string name="join_event_success">Successfully joined the event</string>
+    <string name="join_event_staff">You are already a Staff Member for this Event</string>
```
Comment (Athanaze) [inline app/src/main/res/values/strings.xml:90]: could be shorten to Already a Staff Member for this Event.
Same for the texts afterwards

**Review by Athanaze — APPROVED** [review]

**Review by juaniac — APPROVED** [review]
Good PR overall.
It fixes the issue described neatly. 
It also fixes the tests to work with the updated UI, and adds new tests for the new code.
Thus I approve this PR.

**Pull request merged**